Remove `BaseStore<T>` and replace with composition of dependencies

Nokel81 opened the pull request:
Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Rename ExtensionStore -> BaseExtensionStore
- The name was too close to ExtensionsStore

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Move ExtensionsStore to new format

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Move ClusterStore to new format

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Move UserStore to new format

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Cleanup types to remove multiple cast locations

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Move HotbarStore to new format

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Move WeblinkStore to new format

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Move FileSystemProvisionerStore to new format

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Update snapshots

Signed-off-by: Sebastian Malton <sebastian@malton.name>

ixrock commented on packages/core/src/common/entity-preferences/store.ts:
```diff
@@ -20,30 +20,36 @@ export interface EntityPreferencesStoreModel {
-    this.preferences.replace(data.entities ?? []);
+    runInAction(() => {
+      this.preferences.set(entityId, merge(this.preferences.get(entityId), preferences));
+    });
```
Why do we need to "run in action" that single observable's update? or updating any map/set or single in general observable unit. Afaik `runInAction` needs to optimize update something/optimize reactions/etc? and do it in single "transaction" when updates multiple observables within single function scope, e.g.

```ts
// that would make sense for me for example
runInAction(() => {
      this.preferences.set(entityId, merge(this.preferences.get(entityId), preferences));
      this.somethingObservableToo = "UPDATED!";
    });
```

Nokel81 replied:
Yes `runInAction` is a way to specify a transaction. The reason why updating a single observable doesn't normally need to also be within an action is because we currently have `enforceActions: false` as part of our mobx configuration.

ixrock commented on packages/core/src/extensions/extension-loader/file-system-provisioner-store/file-system-provisioner-store.ts:
```diff
@@ -4,36 +4,49 @@
 }
 
-interface Dependencies extends BaseStoreDependencies {
+interface Dependencies {
```
Why not `extends BaseStoreDependencies` anymore? 
It seems we still pass it as dependencies to `this.store`? If so extending base interface would help with auto-completion when editing these fields.

Nokel81 replied:
Actually we specifically don't pass the whole of `this.dependencies` when we call `this.dependencies.createBaseStore()`.

The reason for this is the point of this PR. It is to make `BaseStore` compostable instead of inheritable. That way if `BaseStore` needs to add or remove a dependency we don't have to add a `di.inject()` call in all the places where subclasses are instantiated but instead only in the `createBaseStoreInejctable`.

ixrock commented on packages/core/src/extensions/extensions-store/extensions-store.ts:
```diff
@@ -23,23 +23,35 @@ export interface IsEnabledExtensionDescriptor {
-  @computed
-  get enabledExtensions() {
-    return Array.from(this.state.values())
+  readonly enabledExtensions = computed(() => (
```
_nit:_
```suggestion
  readonly enabledExtensions: IComputedValue<string[]> = computed(() => (
```

ixrock commented on packages/core/src/main/cluster/manager.ts:
```diff
@@ -246,7 +246,7 @@ export class ClusterManager {
   protected onNetworkOnline = () => {
     this.dependencies.logger.info(`${logPrefix} network is online`);
-    this.dependencies.store.clustersList.forEach((cluster) => {
+    this.dependencies.store.clustersList.get().forEach((cluster) => {
```
_nit:_ why not? 🤪
anyway.. i don't think there is a big need to convert normal observable fields to computed values on the way _(too much)_.

```suggestion
    this.dependencies.get().store.get().clustersList.get().forEach((cluster) => {
```

ixrock reviewed (dismissed):
LGTM

Nokel81 committed:
Clean up impl and rename to better describe intent

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Fix remaining type errors

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Fully split apart the enabled extensions storage

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Fully split apart the clusters storage

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Fully split apart the hotbar storage

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Fully split apart the weblinks storage

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Fully split apart the user preferences storage

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Fix crashing

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Fix tests and snapshots

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Fix integration test failures

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Improve typing to prevent errors in the future.

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Cleanup @k8slens/messaging and friends

- To fix type errors

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Fix lint issue

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Fix type errors

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Fix global override not being complete enough causing tests to fail

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Bump memory for unit tests on CI

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Attempt to fix memory issue on CI again

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Fixup test because of new injectables

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Upgrade Jest

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Fix unit tests falling over

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Back out jest config change

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Remove console log

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Update snapshot

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Fix tests by matching equality instead of snapshots

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Fix tests by forcing specific snapshot style

- Ubuntu CI seems to format arrays in snapshots differently than macOS locally

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Iku-turso reviewed (approved):
Live-reviewed
